pkg/cloud/aws/services/ec2/instances.go

```diff
@@ -376,6 +344,74 @@ func (s *Service) createInstance(machine *actuators.MachineScope, bootstrapToken
 	return out, nil
 }
 
+func setInitConfigurationOptions(initConfig *kubeadmv1beta1.InitConfiguration, machine *clusterv1.Machine) {
+	kubeadm.SetInitConfigurationOptions(
+		initConfig,
+		kubeadm.WithNodeRegistrationOptions(
+			kubeadm.SetNodeRegistrationOptions(
+				&initConfig.NodeRegistration,
+				kubeadm.WithTaints(machine.Spec.Taints),
+				kubeadm.WithNodeRegistrationName(hostnameLookup),
+				kubeadm.WithCRISocket(stringOrDefault(initConfig.NodeRegistration.CRISocket, containerdSocket)),
+				kubeadm.WithKubeletExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+			),
+		),
+	)
+}
+
+func setNodeJoinConfigurationOptions(joinConfig *kubeadmv1beta1.JoinConfiguration, machine *clusterv1.Machine, apiServerEndpoint, bootstrapToken, caCertHash string) {
+	kubeadm.SetJoinConfigurationOptions(
+		joinConfig,
+		kubeadm.WithBootstrapTokenDiscovery(
+			kubeadm.NewBootstrapTokenDiscovery(
+				kubeadm.WithAPIServerEndpoint(apiServerEndpoint),
+				kubeadm.WithToken(bootstrapToken),
+				kubeadm.WithCACertificateHash(caCertHash),
+			),
+		),
+		kubeadm.WithJoinNodeRegistrationOptions(
+			kubeadm.SetNodeRegistrationOptions(
+				&joinConfig.NodeRegistration,
+				kubeadm.WithNodeRegistrationName(hostnameLookup),
+				kubeadm.WithCRISocket(stringOrDefault(joinConfig.NodeRegistration.CRISocket, containerdSocket)),
+				kubeadm.WithKubeletExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+				kubeadm.WithTaints(machine.Spec.Taints),
+				kubeadm.WithKubeletExtraArgs(map[string]string{"node-labels": nodeRole}),
+			),
+		),
+	)
+}
+
+func setControlPlaneJoinConfigurationOptions(joinConfig *kubeadmv1beta1.JoinConfiguration, machine *clusterv1.Machine, apiServerEndpoint, bootstrapToken, caCertHash string) {
+	kubeadm.SetJoinConfigurationOptions(
+		joinConfig,
+		kubeadm.WithBootstrapTokenDiscovery(
+			kubeadm.NewBootstrapTokenDiscovery(
+				kubeadm.WithAPIServerEndpoint(apiServerEndpoint),
+				kubeadm.WithToken(bootstrapToken),
+				kubeadm.WithCACertificateHash(caCertHash),
+			),
+		),
+		kubeadm.WithJoinNodeRegistrationOptions(
+			kubeadm.SetNodeRegistrationOptions(
+				&joinConfig.NodeRegistration,
+				kubeadm.WithTaints(machine.Spec.Taints),
+				kubeadm.WithNodeRegistrationName(hostnameLookup),
+				kubeadm.WithCRISocket(stringOrDefault(joinConfig.NodeRegistration.CRISocket, containerdSocket)),
+				kubeadm.WithKubeletExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+			),
+		),
+		kubeadm.WithLocalAPIEndpointAndPort(localIPV4Lookup, apiServerBindPort),
+	)
+}
+
+func stringOrDefault(val, def string) string {
```

Comment:
This is super generic, what do you think about something like `getCRISocketPath`  which embeds the default in the function? Maybe we can make it generic if we have more stringOrDefaulting to do in future? 

`.WithCRISocket(getCRISocketPath(joinConfig.NodeRegistration.CRISocket))` reads a little nicer to me, but what do you think?

you could even reuse stringOrDefault in the `getCRISocketPath` function if you wanted

Reply:
yeah, I like `WithCRISocket(getCRISocketPath(joinConfig.NodeRegistration.CRISocket))` a lot better. I'll change it.